src/Blocks/BlockGrass.h

```diff
@@ -52,72 +52,102 @@ class cBlockGrassHandler :
-				return;
-			}
 
-			// Make sure that there is enough light at the source block to spread
-			auto light = std::max(a_Chunk.GetBlockLight(AbovePos), a_Chunk.GetTimeAlteredLight(a_Chunk.GetSkyLight(AbovePos)));
-			if (light < 9)
-			{
-				return;
-			}
+		if (!cBlockGrassHandler::TrySurvive(a_Chunk, a_RelPos))
+		{
+			return;
 		}
 
 		// Grass spreads to adjacent dirt blocks:
-		auto & rand = GetRandomProvider();
-		for (int i = 0; i < 2; i++)  // Pick two blocks to grow to
+		for (unsigned i = 0; i < 2; i++)  // Pick two blocks to grow to
 		{
-			int OfsX = rand.RandInt(-1, 1);
-			int OfsY = rand.RandInt(-3, 1);
-			int OfsZ = rand.RandInt(-1, 1);
-
-			BLOCKTYPE  DestBlock;
-			NIBBLETYPE DestMeta;
-			auto Pos = a_RelPos + Vector3i(OfsX, OfsY, OfsZ);
-			if (!cChunkDef::IsValidHeight(Pos.y))
-			{
-				// Y Coord out of range
-				continue;
-			}
-			auto Chunk = a_Chunk.GetRelNeighborChunkAdjustCoords(Pos);
-			if (Chunk == nullptr)
-			{
-				// Unloaded chunk
-				continue;
-			}
-			Chunk->GetBlockTypeMeta(Pos, DestBlock, DestMeta);
-			if ((DestBlock != E_BLOCK_DIRT) || (DestMeta != E_META_DIRT_NORMAL))
-			{
-				// Not a regular dirt block
-				continue;
-			}
-			auto AboveDestPos = Pos.addedY(1);
-			auto AboveBlockType = Chunk->GetBlock(AboveDestPos);
-			auto Light = std::max(Chunk->GetBlockLight(AboveDestPos), Chunk->GetTimeAlteredLight(Chunk->GetSkyLight(AboveDestPos)));
-			if ((Light > 4)  &&
-				cBlockInfo::IsTransparent(AboveBlockType) &&
-				(!IsBlockLava(AboveBlockType)) &&
-				(!IsBlockWaterOrIce(AboveBlockType))
-			)
+			auto & Random = GetRandomProvider();
+			int OfsX = Random.RandInt(-1, 1);
+			int OfsY = Random.RandInt(-3, 1);
+			int OfsZ = Random.RandInt(-1, 1);
+
+			cBlockGrassHandler::TryThrive(a_Chunk, a_RelPos + Vector3i(OfsX, OfsY, OfsZ));
+		}  // for i - repeat twice
+	}
+
+
+
+
+
+	/** Check if conditions are favourable to a grass block at the given position.
+	If they are not, the grass dies and is turned to dirt.
+	Returns whether conditions are so good that the grass can try to spread to neighbours. */
+	static bool TrySurvive(cChunk & a_Chunk, const Vector3i a_RelPos)
+	{
+		const auto AbovePos = a_RelPos.addedY(1);
+		if (!cChunkDef::IsValidHeight(AbovePos.y))
+		{
+			return true;
+		}
+
+		// Grass turns back to dirt when the block Above it is not transparent or water.
+		// It does not turn to dirt when a snow layer is Above.
+		const auto Above = a_Chunk.GetBlock(AbovePos);
+		if (
+			(Above != E_BLOCK_SNOW) &&
+			(!cBlockInfo::IsTransparent(Above) || IsBlockWater(Above)))
+		{
+			a_Chunk.FastSetBlock(a_RelPos, E_BLOCK_DIRT, E_META_DIRT_NORMAL);
+			return false;
+		}
+
+		// Make sure that there is enough light at the source block to spread
+		const auto Light = std::max(a_Chunk.GetBlockLight(AbovePos), a_Chunk.GetSkyLightAltered(AbovePos));
+		return (Light >= 9);
+	}
+
+
+
+
+
+	/** Attempt to spread grass to a block at the given position. */
```

Comment:
This should be `TrySpreadTo`, not obvious enough that RelPos is the pos of the dirtblock rather than the grassblock. Dirt does not 'thrive'!